### DISCUSSION TITLE
dnsdist: is my usage of newServer() correct or there is a better way?

### DISCUSSION THREAD

**Discussion by ousatov-ua** [PowerDNS/pdns#13933] [Q&A] [answered]
Hi!

I have a pdns-recursor in recursion only mode, it runs with next config (I have cpu 16 threads):

```sh
tcp-threads=15
threads=15
```

As facade for it I have running dnsdist with config:

```sh
for i=1,15 do
    addDOHLocal()
    addDOH3Local()
    addTLSLocal()
    addDOQLocal()
    
    # the questions about this downstream:
    newServer({address="127.0.0.1:5353"})
end
```

Is this `newServer()`'s is correct usage or I should play with `sockets` config for newServer() directive?

Sorry if the question is silly, I'm not a clever man.

And do you plan to support downstream `newServer()`'s QUIC?

Thank you in advance!

P.S. Many thanks for contributors for such a great services created!

**Comment by rgacogne** [comment, marked as answer]
It depends on how many CPU cores you want to dedicate to `DNSdist`, mostly. Every single `newServer` directive will start a new thread to deal with the responses from the Recursor, so starting 15 of them can use up to 15 CPU cores. For each directive there is only one socket between dnsdist and the recursor (so one `<destination address, destination port, source address, source port>` tuple).
If you instead use one `newServer` with `sockets=16`, for example, it will start only one thread to deal with incoming responses, but it will use 16 sockets between dnsdist and the Recursor. Increasing the number of sockets might make sense if you are enabling `reuseport` in the recursor configuration, because Linux doesn't load-balance incoming packets very well in that case. We have written more words about that in https://dnsdist.org/advanced/tuning.html so I suggest you start by reading this page, and ask any question you might still have after that.

> And do you plan to support downstream `newServer()`'s QUIC?

If there is actual interest, yes. We have not seen a lot of interest so far, and if I have to guess I suspect people will be more interested in support for outgoing DoH3 than for outgoing DoQ, but I might very well be wrong.

**Comment by ousatov-ua** [comment]
Many thanks for your answer! 
Yes, I see - one thread per one declaration. :)
I will read documentation.


Maybe it is not related to dnsdist (but to Linux itself), but I want to ask you one more question if you don't mind:

In general, I want to utilize all my CPU for dnsdist

How much good dnsdist works in HT, so let's say I have 8 physical core and 16 with HT.
Will it give so much benefit (as 2x) to use exactly 16 instances vs 8 instances of dnsdist?

Thank you!

**Comment by ousatov-ua** [comment]
No needs to answer, I've read documentation.

**Reply by imrankhan-coder** [reply]
> No needs to answer, I've read documentation.

Can you please elaborate it or share the config how you achieved it by utilizing all the CPU threads?
